Add utility functions toposort(), get_node_users(), and use them to optimize universe_getitem after graph parsing

notoraptor opened the pull request:
Hi ! This is a PR to optimise `universe_getitem` nodes after parsing. As I needed `toposort()` and `get_node_users` to do it easily, I implemented these both utility functions here too. Commits are correctly separated, so I could split PR if necessary.

Testing is also added for utility functions and optimization.

@abergeron @breuleux

notoraptor committed:
Move DirectedGraph from Python backend to myia.utils module.
Update DirectedGraph to make it more modular.
- Remove `replace()` method as it is used only in Python backend.
Update Python backend based on new DirectedGraph interface.
- Add a node replacement method for directed graphs, used only here.

notoraptor committed:
Add module myia.ir.graph_utils with 2 functions:
- toposort, using DirectedGraph
- get_node_users, using toposort
Add 2 tests for these functions.

notoraptor committed:
Add post-parsing universe optimization.

Remove useless universe_getitem nodes and related nodes if there is only 1 initial universe_setitem and no other universe_setitem calls in the code.

Uses toposort and get_node_users to optimize.

Add test in tests/test_parser.

notoraptor commented:
PR #406 should be merged and then this one rebased before merging.

breuleux commented on myia/compile/backends/python/optimizer.py:
```diff
@@ -36,15 +36,15 @@ def optimize(self, directed_graphs):
         seen = set()
         while todo:
-            dg = todo.pop(0)
+            dg = todo.pop(0)  # type: DirectedGraph
```
Might as well use an annotation:

```suggestion
            dg: DirectedGraph = todo.pop(0)
```

notoraptor replied:
Ok!

breuleux commented on myia/utils/directed_graph.py:
```diff
@@ -0,0 +1,90 @@
+        # Use a copy to visit graph, so that current graph is not modified.
+        cp = self._copy()
+        # Starts with unused nodes.
+        todo = [n for n in cp.uses if n not in cp.used_by]
```
It would be better to use `collections.deque` instead of a list, the former is more efficient when popping from the left side.

notoraptor replied:
OK!

breuleux commented on myia/ir/graph_utils.py:
```diff
@@ -0,0 +1,68 @@
+    """Return graph nodes in topological order.
+
+    :param graph: myia graph
+    :param reverse: if True, return nodes in execution order (default)
```
I feel the meaning of `reverse` should be the opposite (so it'd be False by default and you'd do `output.reverse()` if reverse is False).

notoraptor replied:
Hi ! By default, as I visit graph from `return_` node to parameters, I get the topological order from user to used nodes, which is not the execution order. As I guess people would like to get execution order, I set `reverse` to True by default, so that one must explicitly set it to `False` to get raw order (user to used), e.g. if execution order is not necessary, or to save time (skipping reversing).

Should I still set `reverse` to False by default ?

Or, for disambiguation, should I rename `reverse` to something like `execution_order` ?

breuleux replied:
I meant that "reverse" being True should mean the opposite order of whatever the default is, so in this case "reverse" would be from user to used, and you'd have to set reverse to True to get the "raw" order. Basically, if a function has a `reverse` argument I expect it to be `False` by default -- the fact that you have to reverse the raw list to get the execution order is an implementation detail, it doesn't have to correspond to the function's API.

notoraptor replied:
OK, updated !

breuleux reviewed (commented):
I wrote a few review comments last week but it appears that I forgot to post them. I haven't gone through everything yet.

notoraptor committed:
In test_code_format:
- test with backend opt deactivated by default
- add test_no_return_parser_opt() with parser opt activated
- backend opt is activated only for test_nonlocal_handle_optimized
In test_parser:
- add test_universe_on_string and test_no_return as cases where universe is still used (no phi used)
- add `parser_opt` versions for these tests. Parser opt removes universe usages.

notoraptor committed:
Use deque instead of list where we need to pop first element from a sequence.

notoraptor commented:
Hi @breuleux @abergeron PR rebased and updated ! I still keep post-parsing opt (added here) and Python backend opt, because there are still cases where phi is not used, so that graph still contains universe methods. I added example cases into `test_parser`.

What do you think ?

notoraptor committed:
Change toposort() API meaning for `reverse` parameter. `topsort` returns nodes execution order by default (reverse=`False`), and raw topological order if reverse is True (from user to used nodes).

abergeron commented on myia/parser.py:
```diff
@@ -77,14 +78,17 @@ def __init__(self, msg, loc):
 
 
-def parse(func):
+def parse(func, parser_opt=False):
```
I'm of the opinion that parser_opt should be always on.

I fail to see a case where one would want to genuinely turn it off.

abergeron commented on myia/parser_opt.py:
```diff
@@ -0,0 +1,90 @@
+        if g in seen_graphs:
+            continue
+        seen_graphs.add(g)
+        g.replace(mapping, mapping_seq, recursive=False)
```
Why are you writing a complex wrapper to do what recursive does?

abergeron commented on myia/parser_opt.py:
```diff
@@ -0,0 +1,90 @@
+                n_typeof = n_make_handle.inputs[0]
+                # Replace universe_getitem nodes with associated value.
+                for n_getitem in nodes_getitem:
+                    _replace_apply_node(graph, n_getitem, n_value)
```
Please use `Graph.replace` directly and add a dictionary to replace all the getitems at once.  With this code you are scanning the graph once per getitem node which is unnecessary and slow.

abergeron commented on myia/parser_opt.py:
```diff
@@ -0,0 +1,90 @@
+    :param new_node: node to replace with. May be None to delete node.
+    """
+    mapping = {node: new_node}
+    mapping_seq = {node: node.edges[SEQ].node} if SEQ in node.edges else {}
```
If you want to delete the SEQ edge, map it to `None`.  An empty dictionary will just use the normal mapping

abergeron commented on myia/parser_opt.py:
```diff
@@ -0,0 +1,90 @@
+                # Remove typeof, universe_setitem and make_handle nodes.
+                _replace_apply_node(graph, n_typeof, None)
+                _replace_apply_node(graph, n_make_handle, None)
+                _replace_apply_node(graph, n_setitem, None)
```
This is invalid, you can't replace a node by `None`, but it works by accident because there are no actual users except the seq chain for the setitem.

Also it will break the sequence chain unless the setitem is the first element.

What you want to do is something like this:
```
prev = n_setitem.edges.get(SEQ, None)
if prev is not None:
    prev = prev.node
graph.replace({}, {n_setitem, prev})
```

abergeron replied:
This could be simplified by a property on Apply nodes similar to `.fn` that returns the next element in the sequence chain or `None` if it's the first, but that is optional for now.

abergeron commented on tests/test_parser.py:
```diff
@@ -229,6 +229,144 @@ def g():
 
 
+def test_self_recursion_parser_opt():
+    def f():
```
For coverage mark this function with `# pragma: no cover` (and all the other tests functions)

abergeron commented on tests/test_parser.py:
```diff
@@ -229,6 +229,144 @@ def g():
+        )
+
+
+def test_universe_on_string():
```
I'm curious what special properties strings have that integers (or any other value) doesn't.

notoraptor committed:
Move graph replacement logic into Graph.replace().

notoraptor committed:
Generate mapping_seq correctly.

notoraptor committed:
Remove wrong node deletion code from Graph.replace()
Remove universe_setitem just by updating SEQ nodes.

notoraptor committed:
Always apply parser opt.
Update tests.

notoraptor committed:
Reformat code.
